File: selenium_test.js
#!/usr/bin/env node
/**
 * Node.js Selenium test script to access localhost:3000 from Docker container
 */

const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const fs = require('fs');

async function createChromeOptions() {
    const options = new chrome.Options();
    options.addArguments('--no-sandbox');
    options.addArguments('--disable-dev-shm-usage');
    options.addArguments('--disable-gpu');
    options.addArguments('--window-size=1920,1080');
    options.addArguments('--disable-extensions');
    options.addArguments('--disable-plugins');
    options.addArguments('--disable-images');
    return options;
}

async function testLocalhostAccess() {
    console.log('🚀 Starting Selenium test...');
    
    const webdriverUrl = 'http://localhost:4444/wd/hub';
    const targetUrl = 'http://host.docker.internal:3000';
    
    let driver;
    
    try {
        // Create Chrome options
        const chromeOptions = await createChromeOptions();
        
        // Connect to remote WebDriver
        console.log(`Connecting to WebDriver at ${webdriverUrl}...`);
        driver = await new Builder()
            .forBrowser('chrome')
            .setChromeOptions(chromeOptions)
            .usingServer(webdriverUrl)
            .build();
        
        console.log(`Navigating to ${targetUrl}...`);
        await driver.get(targetUrl);
        
        // Wait for page to load
        await driver.wait(until.titleIs(await driver.getTitle()), 10000);
        
        // Get page title
        const title = await driver.getTitle();
        console.log('✅ Page loaded successfully!');
        console.log(`Page title: ${title}`);
        
        // Take a screenshot
        const screenshot = await driver.takeScreenshot();
        const screenshotPath = '/tmp/selenium_test_screenshot.png';
        fs.writeFileSync(screenshotPath, screenshot, 'base64');
        console.log(`Screenshot saved to: ${screenshotPath}`);
        
        // Try to find some common elements
        try {
            const body = await driver.findElement(By.tagName('body'));
            const bodyText = await body.getText();
            console.log(`Body content length: ${bodyText.length} characters`);
        } catch (error) {
            console.log(`Could not find body element: ${error.message}`);
        }
        
        console.log('✅ Test completed successfully!');
        return true;
        
    } catch (error) {
        console.log(`❌ Test failed: ${error.message}`);
        return false;
        
    } finally {
        if (driver) {
            await driver.quit();
            console.log('WebDriver session closed');
        }
    }
}

async function checkWebDriverStatus() {
    try {
        const response = await fetch('http://localhost:4444/wd/hub/status');
        if (response.ok) {
            console.log('✅ WebDriver is running and accessible');
            return true;
        } else {
            console.log('❌ WebDriver is not responding properly');
            return false;
        }
    } catch (error) {
        console.log(`❌ Cannot connect to WebDriver: ${error.message}`);
        console.log('Make sure Docker container is running: docker-compose up -d');
        return false;
    }
}

async function main() {
    console.log('🔍 Selenium Test: Accessing localhost:3000 from Docker container');
    console.log('='.repeat(70));
    
    // Check if WebDriver is accessible
    const webdriverOk = await checkWebDriverStatus();
    if (!webdriverOk) {
        process.exit(1);
    }
    
    // Run the test
    const success = await testLocalhostAccess();
    
    if (success) {
        console.log('\n🎉 SUCCESS! Your setup is working correctly!');
        console.log('You can now run Selenium tests against your localhost:3000 service.');
    } else {
        console.log('\n💡 Troubleshooting:');
        console.log('1. Ensure your dev server is running on localhost:3000');
        console.log('2. Check Docker container: docker-compose ps');
        console.log('3. View container logs: docker-compose logs selenium-chrome');
        console.log('4. Access noVNC desktop: http://localhost:7900');
        process.exit(1);
    }
}

// Run the main function
main().catch(console.error);

File: selenium_test.py
#!/usr/bin/env python3
"""
Selenium test script to access localhost:3000 from Docker container
"""

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import sys

def create_chrome_options():
    """Create Chrome options for Docker environment"""
    options = Options()
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    options.add_argument('--disable-gpu')
    options.add_argument('--window-size=1920,1080')
    options.add_argument('--disable-extensions')
    options.add_argument('--disable-plugins')
    options.add_argument('--disable-images')
    return options

def test_localhost_access():
    """Test accessing localhost:3000 from Selenium container"""
    print("🚀 Starting Selenium test...")
    
    # WebDriver configuration
    webdriver_url = "http://localhost:4444/wd/hub"
    target_url = "http://host.docker.internal:3000"
    
    driver = None
    try:
        # Create Chrome options
        chrome_options = create_chrome_options()
        
        # Connect to remote WebDriver
        print(f"Connecting to WebDriver at {webdriver_url}...")
        driver = webdriver.Remote(
            command_executor=webdriver_url,
            options=chrome_options
        )
        
        print(f"Navigating to {target_url}...")
        driver.get(target_url)
        
        # Wait for page to load
        wait = WebDriverWait(driver, 10)
        
        # Get page title
        title = driver.title
        print(f"✅ Page loaded successfully!")
        print(f"Page title: {title}")
        
        # Take a screenshot
        screenshot_path = "/tmp/selenium_test_screenshot.png"
        driver.save_screenshot(screenshot_path)
        print(f"Screenshot saved to: {screenshot_path}")
        
        # Try to find some common elements
        try:
            # Look for common HTML elements
            body = driver.find_element(By.TAG_NAME, "body")
            print(f"Body content length: {len(body.text)} characters")
        except Exception as e:
            print(f"Could not find body element: {e}")
        
        print("✅ Test completed successfully!")
        return True
        
    except Exception as e:
        print(f"❌ Test failed: {e}")
        return False
        
    finally:
        if driver:
            driver.quit()
            print("WebDriver session closed")

def main():
    print("🔍 Selenium Test: Accessing localhost:3000 from Docker container")
    print("=" * 70)
    
    # Check if WebDriver is accessible
    try:
        import requests
        response = requests.get("http://localhost:4444/wd/hub/status", timeout=5)
        if response.status_code == 200:
            print("✅ WebDriver is running and accessible")
        else:
            print("❌ WebDriver is not responding properly")
            sys.exit(1)
    except Exception as e:
        print(f"❌ Cannot connect to WebDriver: {e}")
        print("Make sure Docker container is running: docker-compose up -d")
        sys.exit(1)
    
    # Run the test
    success = test_localhost_access()
    
    if success:
        print("\n🎉 SUCCESS! Your setup is working correctly!")
        print("You can now run Selenium tests against your localhost:3000 service.")
    else:
        print("\n💡 Troubleshooting:")
        print("1. Ensure your dev server is running on localhost:3000")
        print("2. Check Docker container: docker-compose ps")
        print("3. View container logs: docker-compose logs selenium-chrome")
        print("4. Access noVNC desktop: http://localhost:7900")
        sys.exit(1)

if __name__ == "__main__":
    main()
